### PR TITLE
Docker setup

### LINKED ISSUE #58: Add a Docker image for easier development setup
Having a Docker image that can be spun up using `docker-compose` would make it easier to onboard new contributors. I didn't make one since I have no need for it myself, but help is welcome in adding a Docker setup.

The image should use PHP 7.2 for compatibility with the production setup.

### PR DESCRIPTION
**Pull request by quentincaffeino**
- Added docker-compose with required services
- Added dockerfile with php, apache, node, yarn and debugging capabilities
- Added makefile that makes easier working with docker-compose commands

*Bugsquad edit: This closes #58.*

Where I should add readme on using this docker setup? Sure by default it is easy by running `docker-compose up`, but it has some extra options like debugging which requires to be documented.

### CODE REVIEW

**Comment by QuentinCaffeino** [comment]
I've updated Dockerfile accroding to hadolint comments.
Now it's ready to be merged if this is fine with you.

**Comment by QuentinCaffeino** [comment]
I could also open ports for mysql and add phpmyadmin in docker-compose file, which would give more freedom.
Then the problem we would need to solve is ports, we can't bind same port twice and there is a chance dev has something running on 80/3306.
To solve that I could add [Traefik](https://docs.traefik.io/) to docker-compose and bind servers to domains like these `assets.godotengine.org.localhost`, `mysql.assets.godotengine.org.localhost:3306`, `phpmyadmin.assets.godotengine.org.localhost` so there is no problem with remembering or dealing with taken ports.

**Comment by danilopolani** [comment]
Doesn't Traefik need to be on port 80/8080? Maybe just setting a random port for DB and Host could solve the problem, like 3315 for MySQL and 9090 for `web` service.

Otherwise we could set `docker-compose.yml` to be `docker-compose.example.yml` and when cloning the project you have to copy this file and you can customize the ports (like we do with the `.env` file), adding `docker-compose.yml` to `.gitignore`.

I don't know what's the current state of this project, I think currently he does not have enough time for this (?)

**Comment by QuentinCaffeino** [comment]
> Doesn't Traefik need to be on port 80/8080? Maybe just setting a random port for DB and Host could solve the problem, like 3315 for MySQL and 9090 for web service

In proposed solution - yes (but otherwise no, it can run on any port).

> Maybe just setting a random port for DB and Host could solve the problem, like 3315 for MySQL and 9090 for web service.

IMO This is not a solution. This is why I asked before just straight opening ports. Otherwise we would definitely have collisions with different environments.

> Otherwise we could set docker-compose.yml to be docker-compose.example.yml and when cloning the project you have to copy this file and you can customize the ports (like we do with the .env file), adding docker-compose.yml to .gitignore.

I like this solution. WDYT @Calinou?

**Comment by Calinou** [comment]
Is using Traefik in Docker images common for development purposes? It sounds a bit overkill to me.

**Comment by QuentinCaffeino** [comment]
> Is using Traefik in Docker images common for development purposes? It sounds a bit overkill to me.

I wouldn't say it's common.

**Comment by danilopolani** [comment]
> Yes, unfortunately, I don't have much time to finalize everything required to deploy it. (Note that we don't _need_ a Docker setup to deploy it to production, as the production server uses a good old nginx + php-fpm 7.2 setup.)

I thought that it was not required, but it could be useful for contributors, for example on my computer I don't have PHP installed (especially 7.2) and having a docker-compose or whatever would be nice.
I mean, with Docker we could open some PR to help you with the project

**Comment by QuentinCaffeino** [comment]
I'll look again into it later today, what left is:

- [ ] Fix error with PHP heredoc inside blade template
- [ ] Make current docker-compose file an example configuration
- [ ] Add some docs

Did I forgot something?

**Comment by danilopolani** [comment]
If @Calinou agrees on this solution I think that's everything

**Comment by Calinou** [comment]
@QuentinCaffeino Sounds good to me :slightly_smiling_face:

**Comment by QuentinCaffeino** [comment]
@Calinou is there any chance you could send me list of your php extensions? I'll compare them to those that are in container, it has some extra and some removed by default.

**Comment by Calinou** [comment]
I have a lot of extensions on my local setup (including XDebug and the like), not sure if it's very relevant.

That said, it makes sense for the container to include XDebug since it's intended for development.

**Comment by QuentinCaffeino** [comment]
It already does include xdebug.
More about how to work with remote xdebug here: https://github.com/webdevops/php-docker-boilerplate/blob/master/documentation/DOCKER-INFO.md#xdebug-remote-debugger-phpstorm

**Comment by QuentinCaffeino** [comment]
While finalizing this I remembered that XDebug probably won't work as easy with windows and I have no ability to test or fix it. So it would require to create another issue to adapt this setup for windows usage.

The problem is XDebug needs to know your host machine IP in order to connect to XDebug listener (bc it probably runs inside your IDE). I found [this article](https://nickjanetakis.com/blog/docker-tip-65-get-your-docker-hosts-ip-address-from-in-a-container) which сlaims that docker on Mac/Windows have `host.docker.internal` host. But I'm unable to test this properly and this feature is missing on Linux so I used another solution.

Solution was to find out what's host IP before running docker and pass it as one of the docker env variables. But I'm unsure if it would work on windows.

File: docker/Makefile.docker [hunk 1]
```diff
@@ -0,0 +1,39 @@
+
+XDEBUG_REMOTE_HOST := $(shell hostname -I | grep -o "^\S*")
```
Comment (QuentinCaffeino) [inline docker/Makefile.docker:2]: This is the point, which needs extra testing on windows in order for XDebug to work.

**Comment by QuentinCaffeino** [comment]
Ok, this solution of mine was pretty old and I didn't follow docker progress on this issue. A little dig rn took me to [this PR](https://github.com/moby/moby/pull/40007) which must enable `host.docker.internal` for linux. I'll test that and maybe update the setup.

**Comment by QuentinCaffeino** [comment]
I've tried several things, no luck, seems like this feature is not yet in current stable release.

File: docker/docker-compose.yml.example [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    ports:
+      - 8068:80
+    env_file:
+      - ../.env
+    environment:
+      - PMA_HOST=$DB_HOST
+      - PMA_USER=$DB_USERNAME
+      - PMA_PASSWORD=$DB_PASSWORD
+
+
```
Comment (danilopolani) [inline docker/docker-compose.yml.example:42]: I would remove extra white line (at the beginning of the file too) just to clean it up

File: docker/README.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+### 2. Prepairing docker-compose:
+
+```sh
+cp docker/docker-compose.yml.example docker/docker-compose.yml
+```
+
+If you prefer to use local database you can disable mysql and phpmyadmin by removing them from `docker/docker-compose.yml` and setting proper db connection settings in `.env`.
+
+### 3. Setting XDebug defauts (optional):
```
Comment (danilopolani) [inline docker/README.md:28]: There's a typo for `defaults`

```suggestion
### 3. Setting XDebug defaults (optional):
```

**Comment by danilopolani** [comment]
Beside xDebug (which is optional) what's the status of this? I mean at a first look it seems fine (although I didn't test it yet)

**Comment by quentincaffeino** [comment]
I think this is ready

**Comment by rluders** [comment]
Is there something missing here?
I can help if needed. I'm actually running the project inside a docker container with `php-fpm`, `nginx` and `mariadb`.

**Comment by iustin-nita** [comment]
Is this ready to merge ?

**Comment by Calinou** [comment]
@iustin-nita I need to test it locally and see if it everything works as expected before merging it.

**Comment by quentincaffeino** [comment]
Closing because this was superseded by https://github.com/Calinou/godot-asset-library-laravel/pull/415

**Comment by quentincaffeino** [comment]
Closing because this was superseded by https://github.com/Calinou/godot-asset-library-laravel/pull/415

**Pull request closed without merging**